File: julia/nucleotide-count/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
    result = Dict('A' => 0, 'C' => 0, 'G' => 0, 'T' => 0)
    for nucleotide in strand
        if haskey(result, nucleotide)
            result[nucleotide] += 1
        else
            throw(DomainError(nucleotide, "invalid nucleotide"))
        end
    end
    return result
end

File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def ^:private nucleotide? (set "ACGT"))
(def ^:private blank (zipmap nucleotide? (repeat 0)))

(defn count-of-nucleotide-in-strand
  "Given a nucleotide and a string representing a DNA sequence, count how many of the given nucleotide is present.
   If the string contains characters that aren't A, C, G, or T then it is invalid and it throws an assertion error."
  [nucleotide strand]
  (assert (nucleotide? nucleotide))
  (count (filter (partial = nucleotide) strand)))


(defn nucleotide-counts
  "Given a string representing a DNA sequence, count how many of each nucleotide is present.
   If the string contains characters that aren't A, C, G, or T then it is invalid and it throws an assertion error."
  [strand]
  (->> (frequencies strand)
       (merge blank)))
